### DISCUSSION TITLE
Delayed Dispenser Closing

### DISCUSSION THREAD

**Discussion by Jpja** [CounterpartyXCP/CIPs#120] [CIPs]
It was suggested on Telegram that closing of dispensers should be delayed by five blocks. This to prevent an attack vector ("rugspenser") where seller detects incoming dispense and immediately closes the dispenser. This attack is achieved by using a higher fee-ratio for closing the dispenser than was used for the incoming dispense. The closing transaction ends up earlier in the block (by default miner policy) and counterparty therefore closes the dispenser before the dispense is executed. The buyer loses his bitcoin.

The proposed upgrade will **not solve the problem**.

1) the same attack can be pulled off by buying from own dispenser. Instead of setting up his script to _close_ the dispenser when incoming dispense detected, initiate a competing dispense with a higher fee. This is done by sending bitcoin to dispenser from another address. You don't even need a node or rely on Counterparty infrastructure. Set up a python script up against any bitcoin api. It's an easy scam to pull off.

2) delaying closing might create other problems. Are there not legitimate reasons for wanting to close a dispenser immediately? If a delay is added, people might start to close their dispensers by buying from themselves instead (which is instant). It adds complexity for no gain imo.

3) the underlying logic of dispensers is that **the buyer must trust the seller**. Dispenses fail for several reasons
- two or more legitimate buyers at the same time
- seller decides to close at the same time you're buying
- rugspenser attack
- buying from non-counterparty address (eg taproot or bech32-multisig)
- _more?_
The proposed solution only adds some convenience in case of _accidental_ closing, in which case an honest seller will either refund the BTC or send the token. In case of a _vicious_ closing, the attacker will (as explained) switch to frontrunning by buying from own dispenser.

I think the only real solution is twofold:

1) Inform. New users must know that dispenser is an atypical crypto contract. The buyer must be aware that a dispense can fail (accidentally or maliciously) and that he must be **trust** the seller's refund policy.

2) Update the core code. Allow the buyer to send an optional dust transaction to reserve a dispenser (one lot, not the entire dispenser). Let the reservation be active for 10 blocks. Buyer is guaranteed to get the token if the send is confirmed within the deadline. This will make dispensers completely risk free. Personally I never buy from unknown dispensers >₿0.01. With this upgrade I could buy even the most expensive token from a complete stranger – with peace of mind.

**Comment by davestaxcp** [comment]
INFO:

To further the discussion we had here: https://forums.counterparty.io/t/solutions-to-rugspenser-problem/6623

`Inform. New users must know that dispenser is an atypical crypto contract. The buyer must be aware that a dispense can fail (accidentally or maliciously) and that he must be trust the seller's refund policy.`

For any dispenser update. the information and education for both users is the most important part. To truly understand the risks of a function is informed consent and should be shown in the wallet that supports Counterparty dispenser protocol as well as the Github Counterparty Documentation (which is shown on the "Docs" section of Counterparty.io).

Whenever a new feature or process for Dispensers is added, I find it absolutely important to be updating the common user side documentation (website and wallet UI) to inform the user of the risks and process of how this function will happen.

---

RESERVED DISPENSERS:

There is one aspect of this Dispenser upgrade process I think should be considered that I think will solve these issues:

```
two or more legitimate buyers at the same time

buying from non-counterparty address (eg taproot or bech32-multisig)
```

And that is having an optional setting in the wallet UI (and protocol) to have the seller set an EXACT address that is allowed to buy from the dispenser. 

Let's call them "Reserved Dispensers", which in essence is simply a Dispenser that can ONLY be bought or triggered by a specific address the seller chooses to pick that can fill the Dispenser.

If this solution is also added with the "10 blocks wait period" This would not only disable the seller from closing the dispense early (rugging) but also disable the seller from buying the dispenser from themselves while a tx is unconfirmed to buy it.

This is would mean the seller has to already be in communication with the buyer as to which address they will be buying from and then give the buyer a time window for their purchase to happen in a safe manner from ONLY the buyers specific address.

In my opinion, this idea can set up some amazing solutions and applications built around "Reserved Dispensers" as well as making the OTC process of trading XCP assets a little easier because the seller can simply say "I have reserved this card for you, and you will have 10 blocks to buy it safely... and yes only YOUR address will be able to buy it."

**Comment by keyuno** [comment]
While dispensers offer convenience, they unfortunately pose significant risks. As a genuine solution, we should completely phase them out and transition to utilizing DEX.

**Reply by Jpja** [reply]
I think the dispenser is tailored very well for BTC-token trading. The problem is that it is currently only suitable for the green area.
![image](https://github.com/CounterpartyXCP/cips/assets/8447505/82978283-958e-47a6-bf14-2677dc92c7fb)

By adding reservations as an option, the entire graph turns green. The dispenser becomes a **general solution** for BTC-token trading. I'd rather consider phasing out BTC-trading on the DEX. Full reasoning in the [Dispenser Reservation CIP discussion](https://github.com/CounterpartyXCP/cips/discussions/121#discussioncomment-7173037).

In addition to DEX and Dispenser, we need [Atomic Swap](https://github.com/CounterpartyXCP/cips/discussions/100). The former two are not very good when off-chain negotiations are to be settled. Imagine an auction where Alice wins and Bob comes second. With DEX and dispenser, the seller cannot specify who the buyer is, and Bob can frontrun Alice to the trade. Not participating in the auction can therefore be an auction strategy in itself.

This answers @davestaxcp too. I think it complicates dispensers too much to allow the seller specify the buyer. Better add this to Atomic Swap (which seems to be possible with bitcoin scripting only - no Counteparty update needed - with the additional benefit that asset ownerships can be traded risk-free alongside tokens).

**Reply by davestaxcp** [reply]
Thanks for the thoughts JPJA

@keyuno I know you mentioned [here](https://github.com/CounterpartyXCP/cips/discussions/100#discussioncomment-7002703) you were testing this functionality more, but with @Jpja 's comments I think it's a HUGE future CIP that I would love to see integrated :)

It would help community trust and new user trust to have these new features enabled and supported by wallets

**Reply by keyuno** [reply]
> It would help community trust and new user trust to have these new features enabled and supported by wallets

I've delivered on my commitment to extend atomic swaps functionality within the Counterparty ecosystem by introducing [auctions](https://github.com/keyuno/bitcoin-agent). At this stage, what is required from the community is to dive deeper into the code and test it.

**Comment by adamkrellenstein** [comment]
> 2\. Update the core code. Allow the buyer to send an optional dust transaction to reserve a dispenser (one lot, not the entire dispenser). Let the reservation be active for 10 blocks. Buyer is guaranteed to get the token if the send is confirmed within the deadline. This will make dispensers completely risk free. Personally I never buy from unknown dispensers >₿0.01. With this upgrade I could buy even the most expensive token from a complete stranger – with peace of mind.

This is just a reimplementation of BTCPay. The whole reason that BTCPay exists, and is slow, is because you can't do something like dispensers trustlessly. Anyway, yes, the solution is atomic swaps; the delay is just complexity and overhead.
